### PR TITLE
Add custom element option to WindowScroller

### PR DESCRIPTION
**Pull request by andrewbranch**
In reference to #474 

Adds an optional prop `scrollElement` to WindowScroller which, if provided, attaches scroll handlers to the given element instead of to `window`.

@bvaughn I “tested” this via adding an option to the demo (although maybe you’d like to change the way I did it—added a couple things to Application’s `context`—I’m not familiar with HashRouter, but maybe it can be done through props instead?) and it seems to work nicely. I’m happy to help add tests and docs, but wanted to let you take a look at it as soon as possible, and also get your feedback on the best way to add some tests for this. It almost seems like a big group of the tests could just be run twice, once for `window` and once for some custom element.

There are a couple things I wasn’t sure about, so I’ll start adding a few inline comments to ask about specific things.

### CODE REVIEW

File: source/WindowScroller/WindowScroller.css [hunk 1]
```diff
@@ -0,0 +1,3 @@
+.disabled > * {
+  pointer-events: none !important;
```
Comment (andrewbranch) [inline source/WindowScroller/WindowScroller.css]: I’m using this instead of inline styles because a custom scroll container may have many children that need `pointer-events: none`, whereas scrolling the window only needs to apply that style to `body`. So, I wasn’t sure whether to add `!important` or not. It ought to work pretty similarly to the way it used to (replacing inline styles), and it is still overrideable (higher-specificity selectors with `!important`, or inline styles with `!important`). On the other hand, I hate it when libraries use `!important`.

File: source/WindowScroller/WindowScroller.example.js [hunk 2]
```diff
@@ -9,19 +9,27 @@ import AutoSizer from '../AutoSizer'
 import shallowCompare from 'react-addons-shallow-compare'
 import styles from './WindowScroller.example.css'
 
-export default class AutoSizerExample extends Component {
+export default class WindowScrollerExample extends Component {
   static contextTypes = {
-    list: PropTypes.instanceOf(Immutable.List).isRequired
+    list: PropTypes.instanceOf(Immutable.List).isRequired,
+    customElement: PropTypes.object,
```
Comment (andrewbranch) [inline source/WindowScroller/WindowScroller.example.js]: In an earlier commit, I had `PropTypes.instanceOf(Element)`, but the linter complained about that, and I realized that it’s probably for good reason: `Element` won’t be defined on servers, and this is a static property. So, not sure if there’s something better than `object` here? I guess we could make a custom validator function which, when invoked, looks at `window.Element` but gracefully does nothing on server-side render?

File: source/WindowScroller/utils/dimensions.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    : element.getBoundingClientRect().height
+}
+
+/**
+ * Gets the vertical position of an element within its scroll container.
+ * Elements that have been “scrolled past” return negative values.
+ * Handles edge-case where a user is navigating back (history) from an already-scrolled page.
+ * In this case the body’s top position will be a negative number and this element’s top will be increased (by that amount).
+ */
+export function getPositionFromTop (element, container) {
```
Comment (andrewbranch) [inline source/WindowScroller/utils/dimensions.js]: I was having trouble wrapping my mind around this (previously [here](https://github.com/bvaughn/react-virtualized/pull/481/files#diff-668ee30b9b9d3323692d842bec7a5f55L106)), so extra care in making sure this satisfies the original intention is appreciated.

File: source/demo/ContentBox.css [hunk 4]
```diff
@@ -1,5 +1,5 @@
 .ContentBox {
-  flex: 1;
+  flex: 1 0 auto;
```
Comment (andrewbranch) [inline source/demo/ContentBox.css]: This is obviously not directly related to the source changes, but something to this effect was required to make my addition to the demo work (a ContentBox was shrinking and scrolling, rather than `.Body`). I spot checked the other demo changes and didn't see anything obviously broken. Removing `overflow: auto` has the same effect, if that is preferable to you.

File: source/WindowScroller/WindowScroller.css (hunk 1 above)
Comment (bvaughn) [inline source/WindowScroller/WindowScroller.css]: Do you actually mean `.disabled *`?

Hm... either way, `> *` is not a very performant selector so I'm pretty reluctant to use it, and definitely not with such a generic class name as `disabled`. RV classes are named so as to (hopefully) not conflict with any other global CSS names.

Also related- RV has also been shifting away from any CSS-based styles (with the exception of `Table`). Basically anything that's functional is defined via an inline style; anything that's presentational (only `Table`) is defined via CSS to make it easier to override in a variety of ways (without requiring use of `!important`).

File: source/WindowScroller/WindowScroller.example.js (hunk 2 above)
Comment (bvaughn) [inline source/WindowScroller/WindowScroller.example.js]: `any` type is probably okay here. It's not great, but even [the official Flow types use `any` for `ref` type](https://github.com/facebook/flow/blob/7aa23dd3899e02c8d3cd7691890b268db913bd6b/lib/react.js#L125).

File: source/WindowScroller/utils/dimensions.js (hunk 3 above)
Comment (bvaughn) [inline source/WindowScroller/utils/dimensions.js]: Okay. When I do a real review, I'll read this one carefully.

File: source/demo/ContentBox.css (hunk 4 above)
Comment (bvaughn) [inline source/demo/ContentBox.css]: No strong feelings here. That's fine. 😄

File: source/WindowScroller/WindowScroller.css (hunk 1 above)
Comment (andrewbranch) [inline source/WindowScroller/WindowScroller.css]: If I understand correctly, the intention is to disable pointer events on the contents of whatever element is scrolling, without disabling them on the scrolling element itself. So if the contents can now be _N_ number of nodes, either we use CSS to select all children, or we iterate over all _N_ nodes to add/remove inline styles.

To the initial question, `.disabled > *` and `.disabled *` should have the same effect, and I didn't expect the latter to be more efficient, but that’s fine.

And as far as the name goes, we can definitely change it—since the demo was using CSS Modules, and I'm using CSS Modules in my own stuff, I totally didn't think about it. Another option would be to dynamically generate/insert a stylesheet, but that would be a lot more magic; harder for developers to track down if they have the need to customize behavior.

**Comment by andrewbranch** [comment]
Tests were failing due to missing `autoprefixer-loader`, which is referenced in webpack.config.umd.js, but is deprecated, and you have the correct postcss-loader with autoprefixer config in other webpack config files, so I'm guessing you intend to update the UMD config soon, but I went ahead and added the simplest possible fix so we can see test results.

**Comment by andrewbranch** [comment]
Back to the CSS issue, I see we definitely have to do something different. I totally wasn't thinking and did `import styles from '../WindowScroller.css'` in that util, which doesn’t work in the built output, because the CSS file isn't copied, and not everyone is using Webpack / something that can import CSS. So, not sure what the best way forward is.

File: source/WindowScroller/utils/onScroll.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
+import styles from '../WindowScroller.css'
```
Comment (bvaughn) [inline source/WindowScroller/utils/onScroll.js]: Not using CSS modules in RV proper

**Comment by bvaughn** [comment]
> Tests were failing due to missing autoprefixer-loader, which is referenced in webpack.config.umd.js, but is deprecated, and you have the correct postcss-loader with autoprefixer config in other webpack config files. I'm guessing you intend to update the UMD config soon, but I went ahead and added the simplest possible fix so we can see test results.

> Back to the CSS issue, I see we definitely have to do something different. I totally wasn't thinking and did import styles from '../WindowScroller.css' in that util, which doesn’t work in the built output, because the CSS file isn't copied, and not everyone is using Webpack / something that can import CSS. So, not sure what the best way forward is.

I'm intentionally not using CSS modules within react-virtualized because I don't want to impose that requirement on lib users. Things started failing b'c you are importing a stylesheet to use it this way and the build isn't setup to. (Static style names are used.)

File: source/WindowScroller/WindowScroller.css (hunk 1 above)
Comment (andrewbranch) [inline source/WindowScroller/WindowScroller.css]: > The first disables pointer events only for immediate children. The latter disables for all descendants.

Well, yeah, but [`pointer-events` is inherited](http://codepen.io/anon/pen/pNWejN), which I guess is what I was counting on there—analogous to WindowScroller in master counting on elements to inherit it from `document.body`.

>I wonder if it would be more appropriate to approach this by passing a prop (eg disablePointerEvents) to Grid telling it to disable pointer events.

That works, but it _is_ different than the existing effect. WindowScroller in current `master` disables pointer-events on `document.body`, e.g. all the stuff inside the scroll container. My CSS (while obviously not being feasible) accomplishes exactly that for arbitrary scroll containers. What you’re suggesting would narrow the scope of what gets `pointer-events: none` down to just the grid/list, not the whole scroll container. That seems like a reasonable trade-off to me, but I don’t have the same prior knowledge as you about how `pointer-events` effects performance of this whole system. If you’re ok with that trade-off, I can move forward with implementing it.

**Comment by andrewbranch** [comment]
@bvaughn don't want to be a bother, but any thoughts on [my last comments](https://github.com/bvaughn/react-virtualized/pull/481#discussion_r89879941)?

File: source/WindowScroller/WindowScroller.css (hunk 1 above)
Comment (bvaughn) [inline source/WindowScroller/WindowScroller.css]: `pointer-evetns` is an Inherited style, yes, but it can be overridden by children. And `Grid` would override, since it sets the property on the `ReactVirtualized__Grid__innerScrollContainer`, which is not the direct child of `WindowScroller` but- at best- a child of its child.

Also using `!important` is kind of a last resort, so I like to avoid it.

**Comment by bvaughn** [comment]
Yup. Just replied.

**Comment by andrewbranch** [comment]
@bvaughn ok so, at least for the moment, I reverted the contentious stuff and only do anything with pointer-events in the default, original case where `window` itself is scrolling. Do you think other changes to the way pointer events works (e.g. passing down as a prop to Grid) is outside the scope of this PR, or does it need to be addressed here?

**Comment by bvaughn** [comment]
> Do you think other changes to the way pointer events works (e.g. passing down as a prop to Grid) is outside the scope of this PR, or does it need to be addressed here?

I don't think it needs to be addressed here.

Unfortunately I noticed a bug while testing this change locally. Steps to reproduce it:

1. Open the `WindowScroller` demo
2. Click the "Use custom element for scrolling" checkbox
3. Scroll down a bit
4. Open another demo (any one)
5. Return to the `WindowScroller` demo

At this point the list will be cut off in the middle, like:
screen shot 2016-12-01 at 8 56 45 pm

Unfortunately it's been a long, rough day here and I don't have the mental energy to dig into this. I don't feel comfortable moving forward with the PR though without it being addressed.

**Comment by andrewbranch** [comment]
No worries, I'll look into it tomorrow! Definitely need to get that fixed. Thanks for reviewing!

**Comment by bvaughn** [comment]
❤️ thank you for understanding.

**Comment by andrewbranch** [comment]
@bvaughn fixed the bug you were seeing—good catch. When WindowScroller’s `componentDidMount` runs, the list has not yet been rendered, so the scroll container, at the time its height was being calculated and cached, had shrunk down to fit its contents. A moment later, the list finished rendering and the scroll container grew to take up all of its available space (the bottom of the window).

Above, I fixed the issue by making sure that scroll container always grows to fill the window, regardless of its contents. However, do you think this is a good case for using `detectElementResize` on the custom scroll containers?

File: source/WindowScroller/WindowScroller.js [hunk 6]
```diff
@@ -43,27 +47,31 @@ export default class WindowScroller extends Component {
...
-      })
+  componentWillReceiveProps (nextProps) {
+    if (nextProps.scrollElement && nextProps.scrollElement !== this.scrollElement) {
+      this._updateDimensions(nextProps.scrollElement)
+      unregisterScrollListener(this, this.scrollElement)
+      registerScrollListener(this, nextProps.scrollElement)
+    } else if (!nextProps.scrollElement && this.scrollElement !== window) {
+      this._updateDimensions(window)
+      unregisterScrollListener(this, this.scrollElement)
+      registerScrollListener(this, window)
```
Comment (bvaughn) [inline source/WindowScroller/WindowScroller.js]: Am I reading this wrong, or could we reduce the complexity of this a bit?

```js
if (nextProps.scrollElement !== this.scrollElement) {
  this._updateDimensions(nextProps.scrollElement || window)
  unregisterScrollListener(this, this.scrollElement)
  registerScrollListener(this, nextProps.scrollElement || window)
}
```

File: source/WindowScroller/WindowScroller.js [hunk 7]
```diff
@@ -92,7 +113,7 @@ export default class WindowScroller extends Component {
   _onResizeWindow (event) {
     const { onResize } = this.props
 
-    const height = window.innerHeight || 0
+    const height = getHeight(this.scrollElement)
```
Comment (bvaughn) [inline source/WindowScroller/WindowScroller.js]: We should probably call `this._updateDimensions()` here instead? That will handle cases where a resize affects top position (due to things like text wrapping) and it will also only call `setState` if the `height` has actually changed (eg not just a horizontal resize).

The `onResize` call should also be moved into this helper method for similar reasons (and to avoid duplication).

```js
_onResizeWindow (event) {
  this._updateDimensions()
}
```

File: source/WindowScroller/WindowScroller.example.js [hunk 8]
```diff
@@ -9,19 +9,27 @@ import AutoSizer from '../AutoSizer'
...
+    this.onChangeCustomElementCheckbox = this.onChangeCustomElementCheckbox.bind(this)
+  }
+
+  onChangeCustomElementCheckbox (event) {
+    this.context.setScrollingCustomElement(event.target.checked)
   }
 
   render () {
-    const { list } = this.context
+    const { list, isScrollingCustomElement, customElement } = this.context
```
Comment (bvaughn) [inline source/WindowScroller/WindowScroller.example.js]: I know this is probably necessary to test the change within the demo site, but I'm not a fan of the complexity this adds to the `WindowScroller.example` or `Application` components. I wonder if it wouldn't be better to leave this particular functionality _out_ of the demo and just cover it via unit tests instead?

Speaking of which 😁 this change set should be accompanied by a couple of new unit tests so others (including myself) don't break it in the future.

File: source/WindowScroller/WindowScroller.js [hunk 9]
```diff
@@ -83,6 +91,19 @@ export default class WindowScroller extends Component {
...
+  _updateDimensions (scrollElement = this.scrollElement) {
+    const { height } = this.state
+
+    this._positionFromTop = getPositionFromTop(ReactDOM.findDOMNode(this), scrollElement)
+
+    const newHeight = getHeight(scrollElement)
+    if (height !== newHeight) {
+      this.setState({
+        height: newHeight
+      })
```
Comment (bvaughn) [inline source/WindowScroller/WindowScroller.js]: This is probably the place to call the `onResize` prop too.

**Comment by bvaughn** [comment]
Nice work on fixing that bug.

I resumed the code review and left some more thoughts. I also merged in `master` and there were a few conflicts which I've fixed.

To save you a little time, I've pushed the merged result here: https://github.com/bvaughn/react-virtualized/tree/andrewbranch-window-scroller-custom-element

File: source/WindowScroller/WindowScroller.example.js (hunk 8 above)
Comment (andrewbranch) [inline source/WindowScroller/WindowScroller.example.js]: Totally. I was hoping to get your thoughts on what/how to test. Was wondering if some of the existing WindowScroller tests could be extracted into a function and be called once for each "mode" (scrolling `window` and scrolling `scrollElement`). Does that make sense or should each new test be written out distinctly? Do you want to duplicate some of the basic assertions to make sure WindowScroller as a whole doesn’t mess up with `scrollElement`, or should the new tests only address the specific added functionality?

File: source/WindowScroller/utils/dimensions.js [hunk 10]
```diff
@@ -26,5 +26,5 @@ export function getHeight (element) {
  */
 export function getPositionFromTop (element, container) {
   const containerElement = container === window ? document.documentElement : container
-  return element.getBoundingClientRect().top - containerElement.getBoundingClientRect().top
+  return element.getBoundingClientRect().top + getVerticalScroll(container) - containerElement.getBoundingClientRect().top
```
Comment (andrewbranch) [inline source/WindowScroller/utils/dimensions.js]: This fixes a bug which can be reproduced in the demo, before this commit, by

1. Check the "use custom element for scrolling" box
2. Press Shift + Tab twice so that the "Hide header text" button is focused
3. Scroll down a ways
4. Press the space bar so as to press "Hide header text"
5. Scroll again

I also made sure that the back-button navigation edge case still works.

Comment (andrewbranch) [inline source/WindowScroller/utils/dimensions.js]: @bvaughn Can you check this out, because it ostensibly fixes bugs, but breaks some tests 😭

Comment (andrewbranch) [inline source/WindowScroller/utils/dimensions.js]: @bvaughn Any chance you could look at this sometime soonish? Too long in PRgatory makes me anxious 😅

**Comment by bvaughn** [comment]
I'm so sorry. Promise I'll try to make time to look this weekend. Please ping me again if you haven't heard back from my by Saturday.

**Comment by bvaughn** [comment]
PS Thanks for keeping on top of this!

**Comment by andrewbranch** [comment]
No worries! I didn't want to bother you until after the holidays. Maintaining popular open source things sounds hard 😁

**Comment by bvaughn** [comment]
That's very understanding of you 😄  Let's get this done this weekend. We'll talk more soon!

**Comment by bvaughn** [comment]
Looks like this branch has diverged from master a bit, unfortunately. There are changes showing up in the diffs that aren't related, and some things (see below) committed to master haven't made it into this branch either. My fault for taking so long to review it but it is what it is.

If I checkout this branch as-is and merge in master, 4 tests fail:
* WindowScroller onScroll should trigger callback when window scrolls
* WindowScroller onScroll should update :scrollTop when window is scrolled
* WindowScroller updatePosition should recalculate the offset from the top when the window resizes
* WindowScroller updatePosition should recalculate the offset from the top if called externally

These tests were added in 2e91ec7 which was made after your branch and did not merged in. I'll take a stab at reconciling them this morning because I know we'd both really like this PR to get merged. 😄

**Comment by bvaughn** [comment]
Merged in 5cc2ccb, after conflicts resolved.

Thanks!

**Comment by bvaughn** [comment]
Okay @andrewbranch. This was just released in [8.10.0](https://github.com/bvaughn/react-virtualized/blob/master/CHANGELOG.md#8100). Thanks again for your work!

If you'd be willing, I'd still like to get [some test coverage for this new feature](https://github.com/bvaughn/react-virtualized/pull/481/files#r90760446) maybe as a follow-up PR. If you can't get to it I'll find some time to add them.

Cheers!

**Comment by andrewbranch** [comment]
🎉 Thanks for all the help! I can probably add some tests in the near-ish future. Cheers!

**Comment by bvaughn** [comment]
That would be great. Thanks! :)
